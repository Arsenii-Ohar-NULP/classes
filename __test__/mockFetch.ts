export const mockFetch = () => {
    global.fetch = jest.fn();
}