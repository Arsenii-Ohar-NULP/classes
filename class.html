<!DOCTYPE html>
<!--[if lt IE 7]>      <html class="no-js lt-ie9 lt-ie8 lt-ie7"> <![endif]-->
<!--[if IE 7]>         <html class="no-js lt-ie9 lt-ie8"> <![endif]-->
<!--[if IE 8]>         <html class="no-js lt-ie9"> <![endif]-->
<!--[if gt IE 8]>      <html class="no-js"> <!--<![endif]-->
<html>
  <head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <title></title>
    <meta name="description" content="" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <link rel="stylesheet" href="class.css" />
  </head>
  <body>
    <!--[if lt IE 7]>
      <p class="browsehappy">
        You are using an <strong>outdated</strong> browser. Please
        <a href="#">upgrade your browser</a> to improve your experience.
      </p>
    <![endif]-->
    <nav class="navbar navbar-expand-lg bg-body-tertiary" data-bs-theme="dark">
      <div class="container-fluid">
        <a class="navbar-brand" href="#">Classes</a>
        <button
          class="navbar-toggler"
          type="button"
          data-bs-toggle="collapse"
          data-bs-target="#navbarNavDropdown"
          aria-controls="navbarNavDropdown"
          aria-expanded="false"
          aria-label="Toggle navigation"
        >
          <span class="navbar-toggler-icon"></span>
        </button>
        <div
          class="collapse navbar-collapse justify-content-between"
          id="navbarNavDropdown"
        >
          <ul class="navbar-nav">
            <li class="nav-item">
              <a class="nav-link active" aria-current="page" href="#">Home</a>
            </li>
          </ul>
          <ul class="navbar-nav">
            <li class="nav-item">
              <a class="nav-link" href="#">
                <img
                  src="images/how-to-watch-john-wick (2).jpg"
                  height="48"
                  width="48"
                  class="rounded-circle"
                />
              </a>
            </li>
          </ul>
        </div>
      </div>
    </nav>
    <div class="container-sw">
      <div class="container bg-dark text-white w-auto rounded-1 pb-2">
        <div class="hstack justify-content-between">
          <div>
            <div class="p-3 pb-0">
              <p class="fs-3">Learn React.js in a month</p>
            </div>
            <div class="px-3">
              <a>
                <p class="fs-5">Arsenii Ohar</p>
              </a>
            </div>
            <div class="px-3">
              <p class="fs-6">You can learn React.js here quickly.</p>
            </div>
          </div>
          <button class="btn btn-primary mx-3">Join</button>
        </div>
      </div>
      <div class="bg-body-tertiary">
        <div class="container my-2">
          <div class="hstack gap-3">
            <img
              src="./images/the-rock-small.jpg"
              class="rounded-circle"
              width="64"
              height="64"
            />
            <div class="card bg-secondary h-100 p-2 rounded-5">
              <div class="text-white p-2">
                <div class="vstack mx-3 justify-content-start">
                  <p class="fs-5 mb-0">
                    The Rock <span class="badge badge-dark">Student</span>
                  </p>
                  <p class="fs-6 mb-0">
                    Hey everyone, it's The Rock, and I'm thrilled to be taking
                    this online course on React.js, because I have a specific
                    goal in mind. I want to build an amazing online course
                    landing website, and I know that React.js is the perfect
                    technology for the job.
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>

        <div class="container my-2">
          <div class="hstack gap-3">
            <img
              src="./images/the-rock-small.jpg"
              class="rounded-circle"
              width="64"
              height="64"
            />
            <div class="card bg-secondary h-100 p-2 rounded-5">
              <div class="text-white p-2">
                <div class="vstack mx-3 justify-content-start">
                  <p class="fs-5 mb-0">
                    The Rock <span class="badge badge-dark">Student</span>
                  </p>
                  <p class="fs-6 mb-0">
                    I'm excited to learn all about React.js, from the
                    fundamentals to the more advanced topics. As The Rock, I
                    know that with the skills and knowledge I'll gain from this
                    course, I'll be able to create an incredible user experience
                    for my online course landing website.
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>

        <div class="container my-2">
          <div class="hstack gap-3">
            <img
              src="./images/the-rock-small.jpg"
              class="rounded-circle"
              width="64"
              height="64"
            />
            <div class="card bg-secondary h-100 p-2 rounded-5">
              <div class="text-white p-2">
                <div class="vstack mx-3 justify-content-start">
                  <p class="fs-5 mb-0">
                    The Rock <span class="badge badge-dark">Student</span>
                  </p>
                  <p class="fs-6 mb-0">
                    As The Rock, I know the importance of creating a strong
                    online presence, and I'm confident that this course will
                    help me achieve my goals. With React.js, I'll be able to
                    create a website that not only looks great, but functions
                    flawlessly too.
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>

        <div class="container my-2">
          <div class="hstack gap-3">
            <img
              src="./images/the-rock-small.jpg"
              class="rounded-circle"
              width="64"
              height="64"
            />
            <div class="card bg-secondary h-100 p-2 rounded-5">
              <div class="text-white p-2">
                <div class="vstack mx-3 justify-content-start">
                  <p class="fs-5 mb-0">
                    The Rock <span class="badge badge-dark">Student</span>
                  </p>
                  <p class="fs-6 mb-0">
                    I'm grateful for the opportunity to learn from the best in
                    the business, and as The Rock, I can't wait to put my new
                    skills to the test. With the knowledge I'll gain from this
                    course, I know that I'll be able to create an online course
                    landing website that truly stands out from the crowd.
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>

    <script src="" async defer></script>
  </body>
</html>
