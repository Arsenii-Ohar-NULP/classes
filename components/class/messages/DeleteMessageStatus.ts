export enum DeleteMessageStatus {
    OK,
    FAILED,
  }