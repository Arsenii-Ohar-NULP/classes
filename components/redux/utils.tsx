import {getAccessToken} from "../account/TokenService";

export const CLASSES_API_URL = `${process.env["NEXT_PUBLIC_API_URL"]}/api/v1`;


export function authHeaders(headers: Headers) {
    const token = getAccessToken()
    if (token) {
        headers.set("authorization", `Bearer ${token}`)
    }

    return headers;
}

