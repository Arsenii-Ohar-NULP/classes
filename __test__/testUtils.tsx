import React, { PropsWithChildren } from 'react';
import { render, renderHook } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import type { PreloadedState } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import type { RootState, AppStore } from 'pages/redux/store';
// As a basic setup, import your same slice reducers
import { authInitialState, authSlice } from 'pages/redux/auth';
import { classSlice, classesInitialState } from 'pages/redux/classes';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = { auth: authInitialState, classes: classesInitialState },
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { auth: authSlice.reducer, classes: classSlice.reducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderHookWithProviders(
  ui,
  {
    preloadedState = { auth: authInitialState, classes: classesInitialState },
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: { auth: authSlice.reducer, classes: classSlice.reducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...renderHook(() => ui(), { wrapper: Wrapper, ...renderOptions }) };
}
