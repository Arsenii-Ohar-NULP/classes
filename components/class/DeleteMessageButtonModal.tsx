import React, { useState } from 'react';
import { Button, Modal } from 'react-bootstrap';
import { socket } from 'components/utils/socket';
import { DeleteMessageStatus } from './DeleteMessageStatus';

export default function DeleteMessageButtonModal({
  messageId,
  onDelete,
  show,
  close,
}: {
  messageId: number;
  onDelete: () => void;
  show: boolean;
  close: () => void;
}) {
  const [isDeleting, setIsDeleting] = useState<boolean>(false);

  async function deleteMessageSocket() {
    setIsDeleting(true);
    const response = await socket.emitWithAck('deleteMessage', messageId);
    if (response.status === DeleteMessageStatus.OK) {
      onDelete();
      close();
    }
    // TODO: Handle stuff here
    setIsDeleting(false);
  }

  return (
    <>
      <Modal show={show} backdrop={true} onHide={close} keyboard={false}>
        <Modal.Header closeButton={!isDeleting}>
          <Modal.Title>Are you sure?</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          The message will be deleted and there is no way to undo this. Are you
          sure?
        </Modal.Body>
        <Modal.Footer>
          <Button
            variant="danger"
            onClick={deleteMessageSocket}
            disabled={isDeleting}
          >
            Delete
          </Button>
          <Button variant="success" onClick={close} disabled={isDeleting}>
            No, take me back
          </Button>
        </Modal.Footer>
      </Modal>
    </>
  );
}
