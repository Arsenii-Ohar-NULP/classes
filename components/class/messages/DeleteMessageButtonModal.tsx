import React, { useState } from 'react';
import { Button, Modal } from 'react-bootstrap';
import { socket } from 'components/utils/socket';
import { DeleteMessageStatus } from './DeleteMessageStatus';
import {useLogout} from "../../login/AuthService";

export default function DeleteMessageButtonModal({
  messageId,
  onDelete,
  show,
  close,
}: {
  messageId: number;
  onDelete: () => void;
  show: boolean;
  close: () => void;
}) {
  const [isDeleting, setIsDeleting] = useState<boolean>(false);
  const logout = useLogout();
  async function deleteMessageSocket() {
    setIsDeleting(true);
    const response = await socket.emitWithAck('deleteMessage', messageId);
    if (response.status === DeleteMessageStatus.OK) {
      onDelete();
      close();
    }
    else if (response.status === DeleteMessageStatus.FAILED){
      logout();
    }
    setIsDeleting(false);
  }

  return (
    <>
      <Modal show={show} backdrop={true} onHide={close} keyboard={false}>
        <Modal.Header closeButton={!isDeleting}>
          <Modal.Title>Are you sure?</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          The message will be deleted and there is no way to undo this. Are you
          sure?
        </Modal.Body>
        <Modal.Footer>
          <Button
            variant="danger"
            onClick={deleteMessageSocket}
            disabled={isDeleting}
          >
            Delete
          </Button>
          <Button variant="success" onClick={close} disabled={isDeleting}>
            No, take me back
          </Button>
        </Modal.Footer>
      </Modal>
    </>
  );
}
