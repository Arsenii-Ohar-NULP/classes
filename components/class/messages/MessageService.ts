import Forbidden from 'components/errors/Forbidden';
import InvalidCredentials from 'components/errors/InvalidCredentials';
import Message from 'components/class/messages/Message';
import { request } from 'components/utils/Service';
import {getAccessToken} from "components/account/TokenService";

const getApiUrl = () => {
  return process.env['NEXT_PUBLIC_API_URL'];
};

const getMessagesHeader = (): HeadersInit => {
  const token = getAccessToken();

  return {
    'Content-Type': 'application/json',
    Authorization: `Bearer ${token}`,
  };
};

const getMessagesEndpoint = (classId: number) => {
  return getApiUrl() + `/api/v1/class/messages/${classId}`;
};

const handleErrors = (response: Response, msgs: object) => {
  if (response.status == 401) {
    throw new InvalidCredentials(msgs[InvalidCredentials.name]);
  }
  if (response.status == 403) {
    throw new Forbidden(msgs[Forbidden.name]);
  }

  throw new Error(msgs[Error.name]);
};

const fetchMessages = async (classId: number): Promise<Response> => {
  const messagesEndpoint = getMessagesEndpoint(classId);
  const messagesHeader = getMessagesHeader();

  return await fetch(messagesEndpoint, {
    method: 'GET',
    headers: messagesHeader,
  });
};

export const getMessages = async ({
  classId,
}: {
  classId: number;
}): Promise<Message[]> => {
  const response = await fetchMessages(classId);

  if (!response.ok) {
    handleErrors(response, {
      InvalidCredentials: 'Please log in to see the messages',
      Forbidden: 'Join this class to see the messages',
      Error: 'Something went wrong, try again later',
    });
  }

  return (await response.json()) as Message[];
};

const saveMessageHeader = () => {
  const token = getAccessToken();
  return {
    'Content-Type': 'application/json',
    Authorization: `Bearer ${token}`,
  };
};


const deleteMessage = async ({ messageId }: { messageId: number }) => {
  const endpointUrl = `${getApiUrl()}/api/v1/class/message/${messageId}`;
  const headers = saveMessageHeader();

  return await fetch(endpointUrl, {
    method: 'DELETE',
    headers,
  });
};

export const removeMessage = async (messageId: number) => {
  return await request({
    fetchFunction: deleteMessage,
    errors: {
      Error: 'Something went wrong, try again later',
      JsonError: "Couldn't fetch JSON data.",
      InvalidCredentials: 'You should be logged in to delete a message',
      Forbidden: "You don't have access to delete a message",
    },
    args: { messageId },
  });
};
