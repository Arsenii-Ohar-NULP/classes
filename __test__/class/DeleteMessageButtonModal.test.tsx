import * as React from 'react';

import DeleteMessageButtonModal from 'components/class/DeleteMessageButtonModal';
import { renderWithProviders } from '__test__/testUtils';
import { fireEvent, screen, waitFor } from '@testing-library/react';
import { removeMessage } from 'components/class/MessageService';
import { socket } from 'components/utils/socket';
import { DeleteMessageStatus } from 'components/class/DeleteMessageStatus';

jest.mock('components/class/MessageService')
jest.mock('components/login/AuthService')
jest.mock('components/utils/socket', () => ({
  socket: ({
    emitWithAck: jest.fn()
  })
}))

const pushMock = jest.fn((url) => console.log(url));
jest.mock('next/router', () => ({
  useRouter: () => ({
    pathname: '/',
    push: pushMock,
    query: { id: '1' },
  }),
}));

describe('delete message button modal', () => {
  it('should match a snapshot', () => {
    const page = renderWithProviders(
      <DeleteMessageButtonModal
        messageId={1}
        onDelete={jest.fn()}
        show={true}
        close={jest.fn()}
      />
    );
    expect(page).toMatchSnapshot();
  });

  it('should call onDelete, when delete button is clicked', async () => {
    const onDelete = jest.fn();
    const close =  jest.fn();
    jest.mocked(removeMessage).mockResolvedValue({msg: 'Good'})
    jest.mocked(socket.emitWithAck).mockResolvedValueOnce({status: DeleteMessageStatus.OK });
    renderWithProviders(
      <DeleteMessageButtonModal
        messageId={1}
        onDelete={onDelete}
        show={true}
        close={close}
      />
    );

    const button = await screen.findByRole('button', {name: 'Delete'});
    fireEvent.click(button);

    await waitFor(() => {
      expect(jest.mocked(socket.emitWithAck)).toHaveBeenCalled();
      expect(onDelete).toHaveBeenCalled();
      expect(close).toHaveBeenCalled();
    })
  })

  // it('should call logout, when delete button is clicked and invalid credentials is received', async () => {
  //   const onDelete = jest.fn();
  //   jest.mocked(removeMessage).mockRejectedValue(new InvalidCredentials('You should be logged in'));
  //   jest.mocked(socket.emitWithAck).mockResolvedValue(DeleteMessageStatus.FAILED);
  //   renderWithProviders(
  //     <DeleteMessageButtonModal
  //       messageId={1}
  //       onDelete={onDelete}
  //       show={true}
  //       close={jest.fn()}
  //     />
  //   );

  //   const button = await screen.findByRole('button', {name: 'Delete'});
  //   fireEvent.click(button);

  //   await waitFor(() => {
  //     expect(removeMessage).toHaveBeenCalled();
  //     expect(logout).toHaveBeenCalled();      
  //   })
  // })
});
