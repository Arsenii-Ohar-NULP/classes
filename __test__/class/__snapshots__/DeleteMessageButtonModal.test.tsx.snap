// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delete message button modal should match a snapshot 1`] = `<div />`;
